### PR TITLE
forwarder: Use TLS 1.2 by default

### PR DESCRIPTION
**Pull request by Kaderinho**
### What does this PR do?

This PR sets the default value of `min_tls_version` to `tlsv1.2` and completely removes `force_tls_12` from the code logic.

### Motivation

TLS 1.0 and 1.1 are insecure so we should use TLS 1.2 by default. Customers still can use TLS 1.0 and 1.1 by setting `min_tls_version`

### Additional Notes

Can we confirm that we are depreciating the `force_tls_v12` option ? If yes, can I just remove all the references to `force_tls_v12` in the code. (by removing this line `config.BindEnvAndSetDefault("force_tls_12", false)`) ?

### Possible Drawbacks / Trade-offs

That might breaks some customers' configuration if they are sending data to a proxy / non Datadog endpoint which do not support TLS 1.2

### Describe how to test/QA your changes

Changes are unit tested but it is still worth to check if `min_tls_version` is correctly setup.

#### Python script

This scripts creates  several fake TLS servers with different TLS version starting from TLS 1.0 to TLS 1.2. The servers will refuse a connection if the client (Agent in this case) is using a version above the one used by the server.

For instance, if the Agent can connect to a TLS 1.0 server, it means that `min_tls_server` is set to `tlsv1.0` because it would have failed if the `min_tls_server` was `tlsv1.1`.

```python3
# ssl_server.py
import socket
import ssl

def test_conn_tls_version(protocol):

    # This function simply creates a listening socket with a given tls protocol version
    # and waits for a connection. Then it outputs if the connection was successful or not
    # so it's possible to know if the client (Agent in our case) is using the same tls version
    # or above.
    context = ssl.SSLContext(protocol)
    context.load_cert_chain(certfile='cert.pem', keyfile='private.key')

    with socket.socket(socket.AF_INET, socket.SOCK_STREAM, 0) as sock:
        sock.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
        sock.bind(('127.0.0.1', 8443))
        sock.listen(5)

        with context.wrap_socket(sock, server_side=True) as ssock:
            try:
                print("Waiting for connection ...")
                conn, addr = ssock.accept()
                print('Used tls version: ', conn.version())
                conn.close()
                return True
            except ssl.SSLError as e:
                print(f"Agent cannot communicate with the server using {str(protocol)}")
            except Exception:
                pass
    print("==============================================================================\n")
    return False


# Testing connection establishment with tls version starting from 1.0 to 1.2.
# The first successful connection is actually the value of min_tls_version in the Agent
tls_versions =  [(ssl.PROTOCOL_TLSv1, 'tlsv1.0'), (ssl.PROTOCOL_TLSv1_1, 'tlsv1.1'), (ssl.PROTOCOL_TLSv1_2, 'tlsv1.2')]
for tls_version, version_str in tls_versions:
    if test_conn_tls_version(tls_version):
        print(f"Agent min_tls_version is set to {version_str}")
        break
```

#### Agent configuration

Technically we want the following behavior : 

```
if min_tls_version is set and valid:
     min = min_tls_version
else
     min = TLSv1.2
```

If we want to be exhaustive we need to test all the combinations  of : 
- `min_tls_version`: not set, invalid, `tlsv1.0`, `tlsv1.1`, `tlsv1.2`, `tlsv1.3`
- `force_tls_12`: not set, `true`, `false`

To make the testing less cumbersome we can set `force_tls_12: true` and test only the different values for `min_tls_version`. (Technically `force_tls_12` is not used anymore in the code so we should have a warning because the key is unknown, but we're never to sure)

Steps to follow : 
- Stop the Agent
- Set DD_DD_URL to `https://localhost:8443` and `min_tls_version` to the tested value 
- Launch the python script with `python3 ssl_server.py`
- Start the Agent
- Verify the output of the script : the last line should be `Agent min_tls_version is set to tlsv<x.y>`

Here are the values that need to be tested and the expected python output :

- [ ] `min_tls_version: "hello"` -> `tlsv1.2`
- [ ] `min_tls_version: ""` -> `tlsv1.2`
- [ ] `min_tls_version: "tlsv1.0"` -> `tlsv1.0`
- [ ] `min_tls_version: "tlsv1.1"` ->  `tlsv1.1`
- [ ] `min_tls_version: "tlsv1.2"` -> `tlsv1.2`
- [ ] `min_tls_version: "tlsv1.3"` -> Agent is not able to communicate with servers

### Reviewer's Checklist

- [ ] If known, an appropriate milestone has been selected; otherwise the `Triage` milestone is set.
- [ ] Use the `major_change` label if your change either has a major impact on the code base, is impacting multiple teams or is changing important well-established internals of the Agent. This label will be use during QA to make sure each team pay extra attention to the changed behavior. For any customer facing change use a releasenote.
- [ ] A [release note](https://github.com/DataDog/datadog-agent/blob/main/docs/dev/contributing.md#reno) has been added or the `changelog/no-changelog` label has been applied.
- [ ] Changed code has automated tests for its functionality.
- [ ] Adequate QA/testing plan information is provided if the `qa/skip-qa` label is not applied.
- [ ] At least one `team/..` label has been applied, indicating the team(s) that should QA this change.
- [ ] If applicable, docs team has been notified or [an issue has been opened on the documentation repo](https://github.com/DataDog/documentation/issues/new).
- [ ] If applicable, the `need-change/operator` and `need-change/helm` labels have been applied.
- [ ] If applicable, the `k8s/<min-version>` label, indicating the lowest Kubernetes version compatible with this feature. 
- [ ] If applicable, the [config template](https://github.com/DataDog/datadog-agent/blob/main/pkg/config/config_template.yaml) has been updated.

### CODE REVIEW

File: releasenotes/notes/use-tls-1.2-by-default-e2983c21056c3359.yaml
Comment (hush-hush) [inline releasenotes/notes/use-tls-1.2-by-default-e2983c21056c3359.yaml:2]: I think this is more an `upgrade` than `enhancements`, WDYT ?

File: releasenotes/notes/use-tls-1.2-by-default-e2983c21056c3359.yaml
Comment (hush-hush) [inline releasenotes/notes/use-tls-1.2-by-default-e2983c21056c3359.yaml:5]:
```suggestion
    The ``min_tls_version`` configuration parameter now default to `tlsv1.2` instead of `tlsv1.0`.
    Set its value to `tlsv1.0` or `tlsv1.1` if a lower version is needed.
```

File: pkg/config/config_template.yaml
Comment (hush-hush) [inline pkg/config/config_template.yaml:83]: I this is ignored in the code we should remove it from the template and config.go.

File: releasenotes/notes/use-tls-1.2-by-default-e2983c21056c3359.yaml
Comment (hush-hush) [inline releasenotes/notes/use-tls-1.2-by-default-e2983c21056c3359.yaml:10]: WDYT ?
```suggestion
    The deprecated ``force_tls_12`` configuration parameter has been removed in favor of ``min_tls_version`` which default to TLS 1.2.
```

**Review by hush-hush — CHANGES_REQUESTED** [review]

File: pkg/config/config.go
Comment (sgnn7) [inline pkg/config/config.go:366]:
```suggestion
	config.BindEnvAndSetDefault("force_tls_12", false) // deprecated
```

File: pkg/config/config_template.yaml
Comment (sgnn7) [reply pkg/config/config_template.yaml:83]: +1 to this on the doc side. We may want to leave it in the config.go but warn that it will be removed.

File: releasenotes/notes/use-tls-1.2-by-default-e2983c21056c3359.yaml
Comment (Kaderinho) [reply releasenotes/notes/use-tls-1.2-by-default-e2983c21056c3359.yaml:2]: Yeah, I agree ! Thanks 🙏

Comment (sgnn7) [reply releasenotes/notes/use-tls-1.2-by-default-e2983c21056c3359.yaml:2]: Hmm - I think this would be an enhancement - we're not really upgrading a piece of the agent but just bumping a setting

File: releasenotes/notes/use-tls-1.2-by-default-e2983c21056c3359.yaml
Comment (sgnn7) [inline releasenotes/notes/use-tls-1.2-by-default-e2983c21056c3359.yaml:4]:
```suggestion
    The ``min_tls_version`` configuration parameter now defaults to ``tlsv1.2`` instead of ``tlsv1.0``.
```

File: releasenotes/notes/use-tls-1.2-by-default-e2983c21056c3359.yaml
Comment (sgnn7) [inline releasenotes/notes/use-tls-1.2-by-default-e2983c21056c3359.yaml:9]:
```suggestion
    The deprecated ``force_tls_12`` configuration parameter has been removed in favor of ``min_tls_version`` which defaults to TLS v1.2.
```

File: pkg/util/http/transport.go
Comment (sgnn7) [inline pkg/util/http/transport.go:76]: We should have a warning that this setting is deprecated and will be removed if we keep it in config.go

File: pkg/util/http/transport.go
Comment (sgnn7) [inline pkg/util/http/transport.go:77]: nit: extra LF not needed here

**Review by sgnn7 — COMMENTED** [review]
@Kaderinho  Just a few nits from my side

File: pkg/config/config_template.yaml
Comment (Kaderinho) [reply pkg/config/config_template.yaml:83]: I completely agree for `config.go` ! 
For the template, I saw that there was some option still marked as deprecated and I believe they are still here because it might have sense to keep them (at least for 1 or 2 releases) for information only. 

It's not a strong opinion at all here, I just wanted to point this out. If you can confirm that we can remove it from the template without any problem I'll be happy to do so !

File: pkg/config/config.go
Comment (Kaderinho) [reply pkg/config/config.go:366]: I think these space are coming from the linter to match the comment below 🤷

File: releasenotes/notes/use-tls-1.2-by-default-e2983c21056c3359.yaml
Comment (Kaderinho) [reply releasenotes/notes/use-tls-1.2-by-default-e2983c21056c3359.yaml:2]: From Reno:

```
Upgrade notes should be rare: only list known/potential breaking changes,
or major behaviorial changes that require user action before the upgrade.
Notes here must include steps that users can follow to 1. know if they're
affected and 2. handle the change gracefully on their end.
```

It's not a major / critical change but actually it fits with the fact that it might requires some users to change their configuration to keep their setup working (so potential breaking changes). Have no strong opinion on this one as the number of customers affected should be extra low.

File: pkg/config/config_template.yaml
Comment (hush-hush) [reply pkg/config/config_template.yaml:83]: I kind of disagree. The configuration template is not a upgrade doc, the CHANGELOG is there for that. The option in the datadog.yaml marked as deprecated are still supported in the code but will be removed at some point. 

Here the setting was not deprecated but removed.

**Comment by Kaderinho** [comment]
Given all the comments I believe that we have two scenarios: 

- Completely remove references to `force_tls_v12` (in `config.go` and `config_template.yaml`)
- Keep references (for a few releases) even if it's not used anymore + add a warning about it

My opinion is that we should be consistent with what we did previously if we have any example of config parameter deletion. Otherwise I feel like the second option (keep) has a little bit more value as it give information to the customers vs just making an option disappear.

Also, we need to choose between `upgrade` and `enhancements`.

WDYT ? @hush-hush  @sgnn7 (please correct me if I'm wrong)

**Comment by hush-hush** [comment]
> Given all the comments I believe that we have two scenarios:
> 
> * Completely remove references to `force_tls_v12` (in `config.go` and `config_template.yaml`)
> * Keep references (for a few releases) even if it's not used anymore + add a warning about it
> 
> My opinion is that we should be consistent with what we did previously if we have any example of config parameter deletion. Otherwise I feel like the second option (keep) has a little bit more value as it give information to the customers vs just making an option disappear.
 
That field was only used when `min_tls_version` was unset or set to an invalid version. Since we now default to this behavior there is no need to keep it. Removing it will not impact users that are using it (it will just be ignored and TLS 1.2 will be used as expected). The real choice is between removing this field and moving to 1.2 or printing a warning saying that we will move to 1.2 in a few releases.

> Also, we need to choose between `upgrade` and `enhancements`.

I think we should use `upgrade` since we're changing the behavior of the agent. Users look at this section to know if they need to take any actions when upgrading. Here they need to if they want to keep using TLS 1.0.

**Review by jtappa — APPROVED** [review]
👍🏻 from Docs team for upgrade notes

**Comment by sgnn7** [comment]
@Kaderinho I think there was a majority leaning towards both removal of the old option (without trailing warning) and leaning to use of `upgrade` field for the release note. I think we can remove the logic around `force_tls_v12` for sure since it's a noop after we bump `min_tls_version` to TLS v1.2 but I would somewhat prefer we have a warning on `force_tls_v12` option use since the users do not need this field in their configs anymore. With that said, there's no need to hold up this PR over minute details so I'm ok with deferring to the team on what path to choose.

**Review by djmitche — APPROVED** [review]

**Comment by hush-hush** [comment]
> @Kaderinho I think there was a majority leaning towards both removal of the old option (without trailing warning) and leaning to use of `upgrade` field for the release note. I think we can remove the logic around `force_tls_v12` for sure since it's a noop after we bump `min_tls_version` to TLS v1.2 but I would somewhat prefer we have a warning on `force_tls_v12` option use since the users do not need this field in their configs anymore. With that said, there's no need to hold up this PR over minute details so I'm ok with deferring to the team on what path to choose.

The agent already log a warning about all unknown key in the configuration at startup.

File: releasenotes/notes/use-tls-1.2-by-default-e2983c21056c3359.yaml
Comment (hush-hush) [inline releasenotes/notes/use-tls-1.2-by-default-e2983c21056c3359.yaml:5]: I think we should put an emphasize  on the main change which is defaulting to TLS 1.2

WDYT ?
```suggestion
    The Agent now default to TLS 1.2 instead of TLS 1.0. The ``force_tls_12`` configuration parameter has been
    removed since it's now the default behavior.
    To continue using TLS 1.0 or 1.1, you must set the ``min_tls_version`` configuration parameter to either
    `tlsv1.0` or `tlsv1.1`.
```

**Review by hush-hush — APPROVED** [review]
I added a quick comment on the release note. Looks good otherwise.!

File: releasenotes/notes/use-tls-1.2-by-default-e2983c21056c3359.yaml
Comment (Kaderinho) [reply releasenotes/notes/use-tls-1.2-by-default-e2983c21056c3359.yaml]: Yes that's better, thanks ! I'm pinging docs to get a final review on this

File: releasenotes/notes/use-tls-1.2-by-default-e2983c21056c3359.yaml
Comment (cswatt) [inline releasenotes/notes/use-tls-1.2-by-default-e2983c21056c3359.yaml:3]:
```suggestion
    The Agent now defaults to TLS 1.2 instead of TLS 1.0. The `force_tls_12` configuration parameter has been
```

File: releasenotes/notes/use-tls-1.2-by-default-e2983c21056c3359.yaml
Comment (cswatt) [inline releasenotes/notes/use-tls-1.2-by-default-e2983c21056c3359.yaml:5]:
```suggestion
    To continue using TLS 1.0 or 1.1, you must set the `min_tls_version` configuration parameter to either
```

**Review by cswatt — APPROVED** [review]
nits on release notes style, approving so as not to block

File: releasenotes/notes/use-tls-1.2-by-default-e2983c21056c3359.yaml
Comment (Kaderinho) [reply releasenotes/notes/use-tls-1.2-by-default-e2983c21056c3359.yaml]: With grep I can see that there are both single and double backticks in the release notes but the proportion of double-backticks is way more important so you're might be right

Comment (djmitche) [reply releasenotes/notes/use-tls-1.2-by-default-e2983c21056c3359.yaml]: The double-backticks are correct.  I think this is the only place DD uses rST instead of Markdown so it's a common, but incorrect, suggested change.

**Review by jtappa — APPROVED** [review]

**Pull request merged**
